### PR TITLE
Fix false DeprecationWarning in `Module.state_dict`

### LINKED ISSUE #75404: Calling state_dict on nested modules falsely raises DeprecationWarning
### 🐛 Describe the bug

```
from torch.nn import Module

m = Module()
m.state_dict()  # this is fine

m.submodule = Module()
m.state_dict()  # this falsely raises a DeprecationWarning
```

Presumably, this should result in false `DeprecationWarning`s being raised in everyone's code who has nested modules (which is pretty much everyone). More context and pointer to why this is happening here: https://github.com/pytorch/pytorch/pull/72780#pullrequestreview-934487712.

### Versions

main branch (I believe this was introduced by #72780)

cc @ezyang @gchanan @zou3519 @albanD @mruberry @jbschlosser @walterddr @kshitij12345

### PR DESCRIPTION
**Pull request by lkct**
Fixes #75404

TODO:
- [x] add tests

### CODE REVIEW

**Comment by albanD** [comment]
I don't remember if this warning made it to 1.11?

**Comment by lkct** [comment]
> I don't remember if this warning made it to 1.11?

No. Only in nightly.

**Comment by albanD** [comment]
Ok!
Wouldn't a simpler fix to simply make sure we pass these arguments as kwargs when we call into state_dict() there?

**Comment by lkct** [comment]
`destination` is still needed and will cause a warning.

**Comment by lkct** [comment]
@albanD Should we enforce that `state_dict` should contain `_metadata` even if it's an empty dict?

**Comment by albanD** [comment]
>  destination is still needed and will cause a warning.

Ho right, but then we can just call into `_state_dict_impl` directly to avoid that?

> Should we enforce that state_dict should contain _metadata even if it's an empty dict?

We should not change the behavior in this context no.
Also given that this is going away, I don't think we need to worry about it too much.

**Comment by lkct** [comment]
>Ho right, but then we can just call into _state_dict_impl directly to avoid that?

Why solutions not taken
1. `_state_dict_impl()` recursively calls itself.
This does not work because we cannot ignore the overrides of `state_dict()` in `Module` subclasses.
2. `_state_dict_impl()` recursively calls `state_dict()` with proper handling of kwargs.
This is the minimal fix against #72780, but there is actually no need to introduce `_state_dict_impl()` if we're considering simplifying the implementation.

---

> We should not change the behavior in this context no.

I'm not trying to change the behavior of `state_dict`, but we must either allow dict w/o `_metadata` or add it to the following. I'm not sure which way makes more sense.
https://github.com/pytorch/pytorch/blob/70cab5ebb1d3aa536f789905a2bfc51d9858d94e/torch/ao/sparsity/sparsifier/utils.py#L40-L44

**Comment by albanD** [comment]
@lkct I would need to sit down and remind myself of what is happening here. Will answer a bit later when I have time.

**Comment by lkct** [comment]
OK.

I'm now choosing to fix to the original behavior and allow missing `_metadata` to resolve the current check failure.

File: torch/nn/modules/module.py [hunk 1]
```diff
@@ -1384,7 +1358,20 @@ def state_dict(self, *args, destination=None, prefix='', keep_vars=False):
                 " and ".join(warn_msg) + " are deprecated. nn.Module.state_dict will not accept them in the future. "
                 "Refer to https://pytorch.org/docs/master/generated/torch.nn.Module.html#torch.nn.Module.state_dict for details.")
 
-        return self._state_dict_impl(destination, prefix, keep_vars)
+        destination._metadata[prefix[:-1]] = local_metadata = dict(version=self._version)
```
Comment (jbschlosser) [inline torch/nn/modules/module.py:1361]: I think accessing `destination._metadata` here will again cause the issue reported in #72778.

File: torch/nn/modules/module.py [hunk 2]
```diff
@@ -1384,7 +1358,20 @@ def state_dict(self, *args, destination=None, prefix='', keep_vars=False):
...
+        self._save_to_state_dict(destination, prefix, keep_vars)
+        for name, module in self._modules.items():
+            if module is not None:
+                sd = module.state_dict(prefix=prefix + name + '.', keep_vars=keep_vars)
```
Comment (jbschlosser) [inline torch/nn/modules/module.py:1365]: I think this is a bit slower and more memory intensive than just mutating `destination` directly. The difference might be negligible, but possibly not for certain types of modules.

**Review by jbschlosser — COMMENTED** [review]
Did a pass through this; I'll suggest a potential alternative solution that fixes #72778 and doesn't spam warnings:
* Don't deprecate `destination` - remove the deprecation logic and warnings.
* Officially support `destination` with the semantics of mutating it to contain the `state_dict()` stuff, including `_metadata`.
* In `state_dict()`, ensure `destination` always has a `_metadata`; add it if not. I think mutating `destination` to add `_metadata` if it's not already present fits fine with the semantics of `destination`.

@lkct @albanD wdyt about this?

**Comment by albanD** [comment]
If it remains optional, I am happy with keeping it. Even though I don't think the user should use it.
But if removing is too painful, that's fine.

**Comment by lkct** [comment]
@jbschlosser 
> remove the deprecation logic and warnings.

Maybe we should still deprecate positional args?

---

> In state_dict(), ensure destination always has a _metadata; add it if not. I think mutating destination to add _metadata if it's not already present fits fine with the semantics of destination.

If the `destination` passed in is a `dict` (allowed by new typing in #73483), setting custom attr will not be possible since it's a built-in type.
```python
>>> a=dict()
>>> setattr(a,'_metadata',dict())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'dict' object has no attribute '_metadata'
```

---

@albanD 
> Even though I don't think the user should use it.

We can still keep some kind of warning in the docs.

**Comment by jbschlosser** [comment]
> Maybe we should still deprecate positional args?

Yes, I think this is good to do still. For the recursive `state_dict()` calls, we should pass `destination=destination` as a kwarg.

> If the `destination` passed in is a `dict` (allowed by new typing in #73483), setting custom attr will not be possible since it's a built-in type.
> 
> ```python
> >>> a=dict()
> >>> setattr(a,'_metadata',dict())
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
> AttributeError: 'dict' object has no attribute '_metadata'
> ```
> 
> EDIT: One possible workaround
> 
> ```python
> if not isinstance(destination, OrderedDict):
>     destination = OrderedDict(destination)
> ```
> 
> However, the returned dict is not guaranteed to be the same object as `destination`.
> 
> Another possibility I can think of is to change the typings in #73483 and enforce `OrderedDict` for `destination`.

Good points. Instead of going forward with mutating `destination` to add `_metadata`, let's just properly check that `_metadata` exists before trying to access it.

TL;DR: I think we should:
* Deprecate positional args
* Pass `destination=destination` for recursive `state_dict()` calls to avoid triggering the positional arg warning
* Not fully deprecate `destination`, but add a warning in the docs
* Check for `_metadata` before trying to access it
* Add a test to `test/test_nn.py` to make sure `destination` can be passed without triggering a warning and will not error out if it doesn't contain `_metadata` (i.e. ensure #72778 is fixed)

File: torch/nn/modules/module.py [hunk 3]
```diff
@@ -1337,37 +1311,36 @@ def state_dict(self, destination: T_destination, prefix: str = ..., keep_vars: b
...
-
-        .. function:: state_dict(*, prefix='', keep_vars=False)
-           :noindex:
-
-        .. function:: state_dict(destination, prefix='', keep_vars=False)
-           :noindex:
+        .. warning::
+            Currently ``state_dict()`` also accpets positional arguments for
+            ``destination``, ``prefix`` and ``keep_vars`` in order. However,
+            this is being deprecated and keywork arguments will be enforced in
```
Comment (jbschlosser) [inline torch/nn/modules/module.py:1326]: typo nit: keywork -> keyword

File: test/test_nn.py [hunk 4]
```diff
@@ -6170,6 +6170,9 @@ def test_state_dict(self):
         self.assertEqual(state_dict['weight'].data_ptr(), l.weight.data_ptr())
         self.assertEqual(state_dict['bias'].data_ptr(), l.bias.data_ptr())
 
+        # Prference https://github.com/pytorch/pytorch/pull/75507#issuecomment-1110291545
```
Comment (jbschlosser) [inline test/test_nn.py:6173]: typo nit: Prference -> Preference

File: torch/nn/modules/module.py [hunk 5]
```diff
@@ -1337,37 +1311,36 @@ def state_dict(self, destination: T_destination, prefix: str = ..., keep_vars: b
...
 
-        This can be called as
-
-        .. function:: state_dict(*, prefix='', keep_vars=False)
-           :noindex:
-
-        .. function:: state_dict(destination, prefix='', keep_vars=False)
-           :noindex:
+        .. warning::
+            Currently ``state_dict()`` also accpets positional arguments for
```
Comment (jbschlosser) [inline torch/nn/modules/module.py:1324]: typo nit: accpets -> accepts

**Review by jbschlosser — COMMENTED** [review]
Thanks for the update! Looks like it addresses everything we agreed on. There's a couple doc typos (marked below) but otherwise LGTM!

File: torch/nn/modules/module.py (hunk 3 above)
Comment (lkct) [reply torch/nn/modules/module.py]: fixed

File: torch/nn/modules/module.py (hunk 5 above)
Comment (lkct) [reply torch/nn/modules/module.py]: fixed

**Review by jbschlosser — APPROVED** [review]
LGTM!
